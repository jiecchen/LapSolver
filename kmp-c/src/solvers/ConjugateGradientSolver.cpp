#include <mkl.h>
#include <cstdio>
#include "ConjugateGradientSolver.h"

ConjugateGradientSolver::ConjugateGradientSolver(VectorOperator *a, VectorOperator *m, int maxIters, double tolerance)
    : a(a), m(m), maxIters(maxIters), tolerance(tolerance), dim(a->getDimension())
{
    if (m && dim != m->getDimension())
        throw "preconditioner-matrix dimension mismatch";

    tmp = aligned_vector<double>(4 * dim);
}

#define dcg_throw(f) throw fprintf(stderr, "Error! " f " failed with code %d\n", rci), rci;

void ConjugateGradientSolver::apply(double *b, double *x)
{
    // Necessary variables for RCI solver
    int rci;
    double norm = tolerance + 1;

    int one_i = 1;
    double one_d = -1.e0;

    aligned_vector<double> scratch(dim);

    // Initialize the solution
    x[0:dim] = 0.0;

    dcg_init(&dim, x, b, &rci, ipar, dpar, tmp.data());
    if (rci != 0)
        dcg_throw("dcg_init");

    ipar[4] = maxIters;
    ipar[10] = (m != nullptr); // 0 => no preconditioner; else => precondition

    dcg_check(&dim, x, b, &rci, ipar, dpar, tmp.data());
    if (rci != 0)
        dcg_throw("dcg_check");

    do
    {
        dcg (&dim, x, b, &rci, ipar, dpar, tmp.data());
        switch (rci)
        {
        case 0: // PCG is done; just quit
            norm = 0.0;
            break;
        case 1: // Apply the operator to temp(0:n), put the result in tmp(n+1:n)
            a->apply(&tmp[0], &tmp[dim]);
            break;
        case 2: // Compute the norm - decide whether to stop early
            a->apply(x, scratch.data());
            daxpy (&dim, &one_d, b, &one_i, scratch.data(), &one_i);
            norm = dnrm2(&dim, scratch.data(), &one_i);
            break;
        case 3: // Apply the preconditioner, if any. Will never reach if no precon.
            m->apply(&tmp[2 * dim], &tmp[3 * dim]);
            break;
        default:
            dcg_throw("dcg");
        }
    }
    while (norm > tolerance);

    dcg_get (&dim, x, b, &rci, ipar, dpar, tmp.data(), &nIter);
}
