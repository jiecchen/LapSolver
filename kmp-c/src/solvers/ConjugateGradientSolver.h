#pragma once
#include "VectorOperator.h"

class ConjugateGradientSolver : public VectorOperator
{
public:
    ConjugateGradientSolver(VectorOperator *a, VectorOperator *m,
                            int maxIters = 1000, double tolerance = 1e-10);

    virtual int getDimension() const
    {
        return dim;
    }

    virtual int getIterations() const
    {
        return nIter;
    }

    virtual void apply(double *x, double *b);

private:
    VectorOperator *a;
    VectorOperator *m;
    int maxIters;
    double tolerance;

    // Space needed on every call
    aligned_vector<double> x;
    aligned_vector<double> tmp;
    int ipar[128];
    double dpar[128];

    int dim;
    int nIter = 0;
};