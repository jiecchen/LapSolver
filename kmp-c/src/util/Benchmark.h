#pragma once
#include <string>
#include <chrono>
#include <stdexcept>
#include <cstdio>

template<typename Func>
class Benchmark
{
public:
    Benchmark(int argc, char **argv, Func f)
        : name(argv[0])
    {
        for (int i = 1; i < argc; ++i)
        {
            std::string flag(argv[i]);
            if (flag == "-v")
                echo = true;
            else if (flag == "-n" && i + 1 < argc)
            {
                try
                {
                    nTrials = std::stoi(argv[i + 1]);
                }
                catch (...)
                {
                    fprintf(stderr, "bad value for flag -n '%s'\n", argv[i + 1]);
                    break;
                }
            }
        }
        runBenchmark(f);
    }

    Benchmark(std::string name, Func f, bool echo = false, int nTrials = 1)
        : name(name), echo(echo), nTrials(nTrials)
    {
        runBenchmark(f);
    }

    ~Benchmark()
    {
        if (echo)
            fprintf(stderr, "%s completed in %dms (on average over %d trials)\n", name.c_str(), avgTime.count(), nTrials);
    }

    void runBenchmark(Func f)
    {
        auto start = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < nTrials; ++i)
            f();
        auto stop = std::chrono::high_resolution_clock::now();
        avgTime = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start) / nTrials;
    }

private:
    std::string name;
    bool echo = false;
    int nTrials = 1;
    std::chrono::milliseconds avgTime;
};

template<typename Func>
Benchmark<Func> make_benchmark(int argc, char **argv, Func f)
{
    return Benchmark<Func>(argc, argv, f);
}
