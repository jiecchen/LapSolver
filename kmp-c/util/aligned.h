#pragma once
#include "aligned_allocator.h"
#include "aligned_types.h"
#include "aligned_opers.h"
