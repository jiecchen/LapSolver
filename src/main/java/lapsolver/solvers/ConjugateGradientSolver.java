/**
 * @file CGSolver.java
 * @author Cyril Zhang <cyril.zhang@yale.edu>
 * @date Tue Jul 1 2014
 *
 * A wrapper for the Apache Commons-provided conjugate gradient solver.
 */

package lapsolver.solvers;

import lapsolver.Graph;
import lapsolver.util.GraphOperators;
import org.apache.commons.math3.linear.*;

public class ConjugateGradientSolver extends Solver {
    public GraphOperators.LaplacianOperator lapOperator;
    public ConjugateGradient conjugateGradient;
    public RealLinearOperator preconditioner;

    public int maxIters;
    public double tolerance;

    public ConjugateGradientSolver(RealLinearOperator preconditioner, int maxIters, double tolerance) {
        this.preconditioner = preconditioner;
        this.maxIters = maxIters;
        this.tolerance = tolerance;
    }

    // don't use a preconditioner
    public ConjugateGradientSolver(int maxIters, double tolerance) {
        this (null, maxIters, tolerance);
    }

    public void init (Graph graph, double[] d) {
        this.graph = graph;
        this.d = d;
        lapOperator = new GraphOperators.LaplacianOperator(graph, d);
        conjugateGradient = new ConjugateGradient(maxIters, tolerance, true);
    }

    public double[] solve (double[] b) {
        RealVector sol = conjugateGradient.solve(lapOperator, preconditioner, new ArrayRealVector(b));
        return sol.toArray();
    }

}
