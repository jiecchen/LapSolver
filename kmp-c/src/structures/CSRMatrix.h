#pragma once
#include "util/aligned.h"
#include "solvers/VectorOperator.h"
#include "mkl.h"

struct CSRMatrix : public VectorOperator
{
    using VectorOperator::apply;
    virtual void apply(double *x, double *y)
    {
        char n = 'n';
        mkl_cspblas_dcsrgemv(&n, &dim, data.data(), cols.data(), rows.data(), x, y);
    }

    virtual int getDimension() const
    {
        return dim;
    }

    aligned_vector<double> data;
    aligned_vector<int> cols;
    aligned_vector<int> rows;

    int dim;
    int nnz;

};


