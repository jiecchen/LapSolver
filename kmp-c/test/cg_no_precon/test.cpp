#include <solvers/ConjugateGradientSolver.h>
#include <util/aligned.h>
#include <util/Benchmark.h>
#include <structures/CSRMatrix.h>
#include "mkl.h"

/*
 * Note: this test is identical to Intel's own cg_st_criteria.
 *
 * It is here to verify that our interface does not change the
 * correct running of Intel's PCG implementation.
 *
 * It should run in 8 iterations and have one negative zero at
 * dimension 6 (index 5)
 */

int main(int argc, char **argv)
{
    atexit(mkl_free_buffers);

    aligned_vector<int> ia = { 1, 5, 8, 10, 12, 15, 17, 18, 19 };
    aligned_vector<int> ja = { 1, 3, 6, 7, 2, 3, 5, 3, 8, 4, 7, 5, 6, 7, 6, 8, 7, 8 };
    aligned_vector<double> da = { 7, 1, 2, 7, -4, 8, 2, 1, 5, 7, 9, 5, 1, 5, -1, 5, 11, 5 };

    // Change to 0-based indexing
    ia.data()[0:ia.size()]--;
    ja.data()[0:ja.size()]--;

    CSRMatrix<RealSym> a(ia, ja, da);
    ConjugateGradientSolver cg(&a);

    // initialize solution via original operator
    double expected_sol[8] = { 1, 0, 1, 0, 1, 0, 1, 0 };
    aligned_vector<double> rhs = a * expected_sol;

    aligned_vector<double> x(8);
    auto bench = make_benchmark(argc, argv, [&] () {
        x = cg.apply(rhs);
    });

    printf("iters = %d\n", cg.getIterations());
    return 0;
}
