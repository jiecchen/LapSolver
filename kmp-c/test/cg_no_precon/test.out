1.000
0.000
1.000
0.000
1.000
-0.000
1.000
0.000
iters = 8
